ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs

```diff
@@ -0,0 +1,372 @@
+
+instance TPraosCrypto c => CanHardFork (CardanoEras c) where
+  hardForkEraTranslation = EraTranslation {
+      translateLedgerState    = PCons translateLedgerStateByronToShelleyWrapper    PNil
+    , translateLedgerView     = PCons translateLedgerViewByronToShelleyWrapper     PNil
+    , translateConsensusState = PCons translateConsensusStateByronToShelleyWrapper PNil
+    }
+
+{-------------------------------------------------------------------------------
+  Translation from Byron to Shelley
+-------------------------------------------------------------------------------}
+
+translateHeaderHashByronToShelley
+  :: forall sc. Crypto sc
+  => HeaderHash ByronBlock
+  -> HeaderHash (ShelleyBlock sc)
+translateHeaderHashByronToShelley =
+      fromRawHash (Proxy @(ShelleyBlock sc))
+    . toRawHash   (Proxy @ByronBlock)
+
+translatePointByronToShelley
+  :: Crypto sc
+  => Point ByronBlock
+  -> Point (ShelleyBlock sc)
+translatePointByronToShelley = \case
+    GenesisPoint   -> GenesisPoint
+    BlockPoint s h -> BlockPoint s (translateHeaderHashByronToShelley h)
+
+translateUTxOByronToShelley :: forall sc. CC.UTxO -> SL.UTxO sc
+translateUTxOByronToShelley (CC.UTxO utxoByron) =
+    SL.UTxO $ Map.fromList
+      [ (txInShelley, txOutShelley)
+      | (txInByron, txOutByron) <- Map.toList utxoByron
+      , let txInShelley  = translateTxIn  $ CC.fromCompactTxIn  txInByron
+            txOutShelley = translateTxOut $ CC.fromCompactTxOut txOutByron
+      ]
+  where
+    translateTxIn :: CC.TxIn -> SL.TxIn sc
+    translateTxIn (CC.TxInUtxo txId idx) =
+      SL.TxIn (translateTxId txId) (fromIntegral idx)
+
+    translateTxOut :: CC.TxOut -> SL.TxOut sc
+    translateTxOut (CC.TxOut addr amount) =
+      SL.TxOut (translateAddr addr) (translateAmount amount)
+
+    -- | We use the same hasing algorithm so we can unwrap and rewrap the
+    -- bytes. We don't care about the type that is hashed, which will differ
+    -- going from Byron to Shelley, we just use the hashes as IDs.
+    translateTxId :: CC.TxId -> SL.TxId sc
+    translateTxId = SL.TxId . Hash.UnsafeHash . Hashing.hashToBytes
+
+    translateAmount :: CC.Lovelace -> SL.Coin
+    translateAmount = SL.Coin . CC.lovelaceToInteger
+
+    translateAddr :: CC.Address -> SL.Addr sc
+    translateAddr = SL.AddrBootstrap . SL.BootstrapAddress
+
+translateLedgerStateByronToShelleyWrapper
+  :: forall sc. Crypto sc
+  => RequiringBoth
+       WrapLedgerConfig
+       (Translate LedgerState)
+       ByronBlock
+       (ShelleyBlock sc)
+translateLedgerStateByronToShelleyWrapper =
+    RequireBoth $ \_ (WrapLedgerConfig cfgShelley) ->
+    Translate   $ \epochNo ledgerByron ->
+      translateLedgerStateByronToShelley cfgShelley epochNo ledgerByron
+
+translateLedgerStateByronToShelley
+  :: forall sc. Crypto sc
+  => LedgerConfig (ShelleyBlock sc)
+  -> EpochNo
+  -> LedgerState ByronBlock
+  -> LedgerState (ShelleyBlock sc)
+translateLedgerStateByronToShelley cfgShelley epochNo ledgerByron =
```

Comment:
@nc6 @JaredCorduan Can you both have a look at all these `translate*` functions? Most of them should move to `cardano-ledger-specs` anyway.

Reply:
https://github.com/input-output-hk/ouroboros-network/issues/2253